### PR TITLE
RtpStreamSend: sanitization

### PR DESCRIPTION
**Pull request by jmillan**

### CODE REVIEW

**Comment by jmillan** [comment]
I fixed some const correctness in the way of assuring we were not modifying the original RTP packet.

**Comment by jmillan** [comment]
I don't consider the branch finished yet. It still needs to be further tested, but I want to add this sanitizations  here.

File: worker/src/RTC/RtpStreamSend.cpp [hunk 1]
```diff
@@ -101,40 +114,34 @@ namespace RTC
...
-			while (!this->buffer.front())
-			{
-				this->buffer.pop_front();
-				this->startSeq++;
-			}
-		}
-		// If we have erased the last element, remove all `nullptr` elements from the end of the buffer.
-		else if (idx == static_cast<uint16_t>(this->buffer.size() - 1))
+		// Remove all `nullptr` elements from the beginning of the buffer.
+		// NOTE: Calling front on an empty container is undefined.
```
Comment (nazar-pc) [inline worker/src/RTC/RtpStreamSend.cpp:139]: Oh, wow, such an awkward API! I would have never thought that calling front can be so dangerous :exploding_head: 

Can totally be a source of weird random issues I have seen.

File: worker/src/RTC/RtpStreamSend.cpp [hunk 2]
```diff
@@ -154,6 +161,23 @@ namespace RTC
 		this->startSeq = 0;
 	}
 
+	void RtpStreamSend::StorageItemBuffer::Dump()
+	{
+		for (size_t i{ 0 }; i < this->buffer.size(); i++)
```
Comment (nazar-pc) [inline worker/src/RTC/RtpStreamSend.cpp:166]: Why not using iterator API instead here?

File: worker/src/RTC/RtpStreamSend.cpp [hunk 3]
```diff
@@ -419,13 +446,13 @@ namespace RTC
 		auto seq          = packet->GetSequenceNumber();
 		auto* storageItem = this->storageItemBuffer.Get(seq);
 
+		this->ClearOldPackets(packet);
```
Comment (nazar-pc) [inline worker/src/RTC/RtpStreamSend.cpp:449]: Why moving out of `if` that is below? Cleanup is not needed unless that check and timestamps check below is `true`

File: worker/src/RTC/RtpStreamSend.cpp [hunk 4]
```diff
@@ -494,8 +481,43 @@ namespace RTC
 
 		// Store original packet and some extra info into the retrieved storage item.
 		storageItem->originalPacket = clonedPacket;
-		storageItem->ssrc           = clonedPacket->GetSsrc();
-		storageItem->sequenceNumber = clonedPacket->GetSequenceNumber();
+		storageItem->ssrc           = packet->GetSsrc();
+		storageItem->sequenceNumber = packet->GetSequenceNumber();
+		storageItem->timestamp      = packet->GetTimestamp();
```
Comment (nazar-pc) [inline worker/src/RTC/RtpStreamSend.cpp:486]: I remember the timestamp is not modified in `v3` unless I recall it incorrectly, why is it necessary now?

File: worker/src/RTC/RtpStreamSend.cpp [hunk 5]
```diff
@@ -494,8 +481,43 @@ namespace RTC
...
+		MS_TRACE();
+
+		auto packetTs{ packet->GetTimestamp() };
+		auto clockRate{ this->params.clockRate };
+
+		const auto bufferSize = this->storageItemBuffer.GetBufferSize();
+
+		// Go through all buffer items starting with the first and free all storage
+		// items that contain packets older than `MaxRetransmissionDelay`.
+		for (size_t i{ 0 }; i < bufferSize && this->storageItemBuffer.GetBufferSize() != 0; ++i)
```
Comment (nazar-pc) [inline worker/src/RTC/RtpStreamSend.cpp:500]:
```suggestion
		// Go through all buffer items starting with the first and free all storage
		// items that contain packets older than `MaxRetransmissionDelay`.
		for (size_t i{ 0 }; i < this->storageItemBuffer.GetBufferSize(); ++i)
```

**Review by nazar-pc — COMMENTED** [review]
It is harder to review when both `const`, moving code around and other changes are combined in the same commit.

File: worker/src/RTC/RtpStreamSend.cpp (hunk 5 above)
Comment (nazar-pc) [reply worker/src/RTC/RtpStreamSend.cpp:500]: This is a wrong suggestion, sorry

**Comment by jmillan** [comment]
> It is harder to review when both const, moving code around and other changes are combined in the same commit.

Yes, it's true.

I'm thinking I would like to dissect your `memory-optimization` branch into different PRs indeed. This is way too big branch to be securely merged at once. 

We can have the branch until the corresponding PRs are done and merged. I can do this.

File: worker/src/RTC/RtpStreamSend.cpp (hunk 2 above)
Comment (jmillan) [reply worker/src/RTC/RtpStreamSend.cpp:166]: Just to be able to print the index if the `StorageItem` is `nullptr.

File: worker/src/RTC/RtpStreamSend.cpp (hunk 1 above)
Comment (jmillan) [reply worker/src/RTC/RtpStreamSend.cpp:139]: yep.., I faced crash here.

File: worker/src/RTC/RtpStreamSend.cpp (hunk 3 above)
Comment (jmillan) [reply worker/src/RTC/RtpStreamSend.cpp:449]: It is, it is: https://github.com/versatica/mediasoup/blob/v3/worker/src/RTC/SimulcastConsumer.cpp#L876

File: worker/src/RTC/RtpStreamSend.cpp (hunk 4 above)
Comment (jmillan) [reply worker/src/RTC/RtpStreamSend.cpp:486]: It is: https://github.com/versatica/mediasoup/blob/v3/worker/src/RTC/SimulcastConsumer.cpp#L876

File: worker/src/RTC/RtpStreamSend.cpp (hunk 3 above)
Comment (jmillan) [reply worker/src/RTC/RtpStreamSend.cpp:449]: I'm just being super cautious. `ClearOldPackets` will just return right away in that case, without much processing.

**Comment by jmillan** [comment]
I've removed the const correctness changes from this PR

File: worker/src/RTC/RtpStreamSend.cpp (hunk 4 above)
Comment (nazar-pc) [reply worker/src/RTC/RtpStreamSend.cpp:486]: Hm, but clearly not for all consumer types: https://github.com/versatica/mediasoup/blob/v3/worker/src/RTC/SimpleConsumer.cpp#L277-L279
Why? Is there a problem with using original timestamps?

File: worker/src/RTC/RtpStreamSend.cpp (hunk 3 above)
Comment (jmillan) [reply worker/src/RTC/RtpStreamSend.cpp:449]: Also, if not done here we would have to call it from two different places. IMHO it's cleaner having it here. We can move it if we find performance penalties.

File: worker/src/RTC/RtpStreamSend.cpp (hunk 4 above)
Comment (jmillan) [reply worker/src/RTC/RtpStreamSend.cpp:486]: here: https://github.com/versatica/mediasoup/pull/302

Comment (nazar-pc) [reply worker/src/RTC/RtpStreamSend.cpp:486]: Thanks. Why is similar change not needed in non-simulcast consumers though?

**Review by nazar-pc — APPROVED** [review]

**Pull request merged**